File: 2017/lib/day_02.ex
defmodule AdventOfCode2017.Day02 do
  def part1(input) do
    input
    |> parse()
    |> Enum.map(&get_row_difference/1)
    |> Enum.sum()
  end

  def part2(input) do
    input
    |> parse()
    |> Enum.map(&get_row_divided/1)
    |> Enum.sum()
  end

  def parse(input) do
    input
    |> String.split("\n")
    |> Enum.map(&String.split/1)
    |> Enum.map(fn s ->
      Enum.map(s, &String.to_integer/1)
    end)
  end

  def get_row_difference(row), do: Enum.max(row) - Enum.min(row)

  def get_row_divided(row), do: hd(for x <- row, y <- row, x != y, rem(x, y) == 0, do: div(x, y))
end

File: 2015-ruby/lib/day_02.rb
class Day02
  def part1(input)
    prepare_dimensions(input).map(&method(:wrapping_paper)).sum
  end

  def part2(input)
    prepare_dimensions(input).map(&method(:ribbon)).sum
  end

  def ribbon((l, w, h))
    options = [2 * l + 2 * w, 2 * w + 2 * h, 2 * l + 2 * h]
    options.min + l * w * h
  end

  def wrapping_paper((l, w, h))
    sides = [l * w, w * h, h * l]
    2 * sides.sum + sides.min
  end

  def prepare_dimensions(input) 
    input.split.map{ |v| v.split("x").map(&:to_i) }
  end
end
